### PR TITLE
Sparse decomposition for  prepare

### PR DESCRIPTION
**Pull request by fdmalone**
Final PR in series (#546  #547) which adds a decomposition for Prepare. I've added a utility function to generate the necessary integral inputs and a factory method to build the alt / keep values from these. I also add back the notebook using  autotest/autodoc.

### CODE REVIEW

**Comment by fdmalone** [comment]
xref #335

File: qualtran/bloqs/chemistry/sparse/prepare.py [hunk 1]
```diff
@@ -13,22 +13,104 @@
...
+    assert len(tpq_prime.shape) == 2, "hcore should be a matrix"
+    assert len(eris.shape) == 4, "eris should be 4-index tensor"
+    num_spat = tpq_prime.shape[-1]
+    tril = np.tril_indices(num_spat)
+    # we don't sparsify one-body, but just take the upper triangular part
+    tpq_sparse = tpq_prime[tril]
+    tpq_indx = np.array([(ix[0], ix[1], 0, 0) for ix in zip(*tril)])
+    eris_eight = []
+    pqrs_indx = []
+    # ignoring scaling factors for the moment.
```
Comment (mpharrigan) [inline qualtran/bloqs/chemistry/sparse/prepare.py:73]:
```python
eris_eight = []
pqrs_indx = []
def _add(p, q, r, s):
  nonlocal eris_eight, pqrs_indx
  eris_eight.append(eris[p,q,r,s])
  pqrs_indx.append((p, q, r, s))

for p, q, r, s in [various]:
  _add(p, p, q, r)
  _add(p, q, p, r)
  ...
```

File: qualtran/bloqs/chemistry/sparse/prepare.py [hunk 2]
```diff
@@ -132,6 +226,205 @@ def junk_registers(self) -> Tuple[SelectionRegister, ...]:
...
+            theta[0],
+            flag_1b,
+            alt_pqrs[0],
+            alt_pqrs[1],
+            alt_pqrs[2],
+            alt_pqrs[3],
+            theta[1],
+            alt_flag_1b,
+            keep,
+        ) = bb.add(
```
Comment (mpharrigan) [inline qualtran/bloqs/chemistry/sparse/prepare.py:358]: style: you could use `bb.add_d` which returns the output soquets as a dictionary and you can save some vertical space

**Review by mpharrigan — APPROVED** [review]

File: qualtran/bloqs/chemistry/sparse/prepare.py (hunk 1 above)
Comment (fdmalone) [reply qualtran/bloqs/chemistry/sparse/prepare.py:73]: Done.

File: qualtran/bloqs/chemistry/sparse/prepare.py (hunk 2 above)
Comment (fdmalone) [reply qualtran/bloqs/chemistry/sparse/prepare.py:358]: I'm not sure, wouldn't I have to do a lot of dictionary unpacking further on, like x=soqs['target12'], ...?

Comment (mpharrigan) [reply qualtran/bloqs/chemistry/sparse/prepare.py:351]: yeah maybe it wouldn't make anything better

**Pull request merged**